backend/src/controllers/utils/index.ts

```diff
@@ -72,3 +80,204 @@ export const getTemplatesPath = (): string => {
+        percent: '0',
+        taxScheme: taxSchemeIDVAT
+      }
+    } else {
+      taxObject = {
+        id: standardTaxChargeCode,
+        percent: '21',
+        taxScheme: taxSchemeIDVAT
+      }
+      classifiedTaxObject = taxObject;
+    }
+
+
+    const taxTotal = new TaxTotal({
+      taxAmount: new UdtAmount(savedInvoice.money.totalTax.toFixed(2), currencyID),
+      taxSubtotals: [new TaxSubtotal({
+        taxableAmount: new UdtAmount(savedInvoice.money.totalWithoutTax.toFixed(2), currencyID),
+        taxAmount: new UdtAmount(savedInvoice.money.totalTax.toFixed(2), currencyID),
+        taxCategory: new TaxCategory(taxObject)
+      })]
+    });
+
+    /** more info on PaymentMeans codes: https://docs.peppol.eu/poacc/billing/3.0/codelist/UNCL4461/ */
+    const debitPaymentMeansCode = '31';
+    const paymentMeans = new PaymentMeans({
+      paymentMeansCode: debitPaymentMeansCode,
+      payeeFinancialAccount: new PayeeFinancialAccount({
+        id: savedInvoice.your.iban,
+        financialInstitutioBranch: new FinancialInstitutionBranch({
+          id: savedInvoice.your.bic
+        })
+      })
+    });
+
+    const orderRef = savedInvoice.orderNr && savedInvoice.orderNr.trim().length !== 0 ? savedInvoice.orderNr : savedInvoice._id;
+    
+    /** More info on invoice type codes: https://docs.peppol.eu/poacc/billing/3.0/codelist/UNCL1001-inv/ 
+     *  Code 380 signals a commercial invoice. 
+    */
+    const commercialInvoiceTypeCode = '380';
+
+    invoiceXml.addProperty('xmlns', 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2');
+    invoiceXml.addProperty('xmlns:cac', 'urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2');
+    invoiceXml.addProperty('xmlns:cbc', 'urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2');
+    invoiceXml.setCustomizationID('urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0');
+    invoiceXml.setProfileID('urn:fdc:peppol.eu:2017:poacc:billing:01:1.0');
+    invoiceXml.setIssueDate(moment(savedInvoice.date).format('YYYY-MM-DD'));
+    invoiceXml.setDueDate(moment(savedInvoice.date).add(30, 'days').format('YYYY-MM-DD'));
+    invoiceXml.setInvoiceTypeCode(commercialInvoiceTypeCode);
+    invoiceXml.setDocumentCurrencyCode(DEFAULT_CURRENCY);
+    invoiceXml.setAccountingSupplierParty(accountingSupplierParty);
+    invoiceXml.setAccountingCustomerParty(accountingCustomerParty);
+    invoiceXml.setLegalMonetaryTotal(legalMonetaryTotal);
+    invoiceXml.setID(savedInvoice.number.toString());
+    invoiceXml.setOrderReference(new OrderReference({id: orderRef}));
+    invoiceXml.addTaxTotal(taxTotal);
+    invoiceXml.addPaymentMeans(paymentMeans);
+
+    /** more info on unit codes: https://docs.peppol.eu/poacc/billing/3.0/codelist/UNECERec20/ 
+     * code C62 is a general code meaning 'one' or 'unit'
+    */
+    const defaultUnitCode = 'C62';
+    savedInvoice.lines.forEach((line, index) => {
```

Comment:
fyi (for your information) -- you can leave this code AS IS

```
savedInvoice.lines.forEach(...)
```

Avoid this kind of construct (unless good reason to do so). A forEach means that there is going to be one or more side effects in the function. You need to read the entire function and hope that you spot all the data-manipulations. This makes it difficult to understand what is going on / going to happen.

Prefer using side-effect free code:
```
const xmlInvoiceLines = savedInvoice.lines.map((line, index) => { ... return new InvoiceLine(...); }
invoiceXml.addInvoiceLines(xmlInvoiceLines)
```

It looks almost the same, but it makes your functions pure and side-effects evident...

Reply:
Thanks for the feedback, I hadn't thought about this! I will keep this in mind in the future